src/wp-includes/media.php

```diff
@@ -5449,3 +5452,282 @@ function wp_increase_content_media_count( $amount = 1 ) {
+	$image_meta = wp_get_attachment_metadata( $attachment->ID );
+	if ( ! is_array( $image_meta ) ) {
+		return $attr;
+	}
+
+	if ( isset( $image_meta['has_transparency'] ) ) {
+		$attr['data-has-transparency'] = $image_meta['has_transparency'] ? 'true' : 'false';
+
+		$class = $image_meta['has_transparency'] ? 'has-transparency' : 'not-transparent';
+		if ( empty( $attr['class'] ) ) {
+			$attr['class'] = $class;
+		} else {
+			$attr['class'] .= ' ' . $class;
+		}
+	}
+
+	if ( ! empty( $image_meta['dominant_color'] ) ) {
+		$attr['data-dominant-color'] = esc_attr( $image_meta['dominant_color'] );
+		if ( empty( $attr['style'] ) ) {
+			$attr['style'] = '';
+		}
+		$attr['style'] .= '--dominant-color: #' . esc_attr( $image_meta['dominant_color'] ) . ';';
+	}
+
+	return $attr;
+}
+
+/**
+ * Filter image tags in content to add the dominant color to the image.
+ *
+ * @since 6.1
+ *
+ * @param string $filtered_image The filtered image.
+ * @param string $context        The context of the image.
+ * @param int    $attachment_id  The attachment ID.
+ * @return string image tag
+ */
+function img_tag_add_dominant_color( $filtered_image, $context, $attachment_id ) {
+
+	// Only apply this in `the_content` for now, since otherwise it can result in duplicate runs due to a problem with full site editing logic.
+	if ( 'the_content' !== $context ) {
+		return $filtered_image;
+	}
+
+	$image_meta = wp_get_attachment_metadata( $attachment_id );
+	if ( ! is_array( $image_meta ) ) {
+		return $filtered_image;
+	}
+
+	/**
+	 * Filters whether dominant color is added to the image.
+	 *
+	 * You can set this to false in order disable adding the dominant color to the image.
+	 *
+	 * @since 6.1
+	 *
+	 * @param bool   $add_dominant_color Whether to add the dominant color to the image. default true.
+	 * @param int    $attachment_id      The image attachment ID.
+	 * @param array  $image_meta         The image meta data all ready set.
+	 * @param string $filtered_image     The filtered image. html including img tag
+	 * @param string $context            The context of the image.
+	 */
+	$check = apply_filters( 'img_tag_add_dominant_color', true, $attachment_id, $image_meta, $filtered_image, $context );
+	if ( ! $check ) {
+		return $filtered_image;
+	}
+
+	$data        = '';
+	$style       = '';
+	$extra_class = '';
+
+	if ( ! empty( $image_meta['dominant_color'] ) ) {
+		$data .= sprintf( 'data-dominant-color="%s" ', esc_attr( $image_meta['dominant_color'] ) );
+		$style = 'style="--dominant-color: #' . esc_attr( $image_meta['dominant_color'] ) . ';" ';
+	}
+
+	if ( isset( $image_meta['has_transparency'] ) ) {
+		$transparency = $image_meta['has_transparency'] ? 'true' : 'false';
+		$data        .= sprintf( 'data-has-transparency="%s" ', $transparency );
+		$extra_class  = $image_meta['has_transparency'] ? 'has-transparency' : 'not-transparent';
+	}
+
+	if ( ! empty( $data ) || ! empty( $style ) ) {
```

Comment:
```suggestion
	if ( ! empty( $data ) || ! empty( $style ) || ! str_contains( $style,  $filtered_image )  || ! str_contains( $data,  $filtered_image ) ) {
```

Should we check if this in the string before adding it?